packages/zoe/src/contractFacet/zcfZygote.js

```diff
@@ -149,6 +165,78 @@ export const makeZCFZygote = async (
     return zcfMintFactory.makeZCFMintInternal(keyword, zoeMint);
   };
 
+  const HandleOfferShape = M.remotable('HandleOffer');
+
+  // handleOfferObject gives Zoe the ability to notify ZCF when a new seat is
+  // added in offer(). ZCF responds with the exitObj and offerResult.
+  const makeHandleOfferObj = vivifyKind(
+    zcfBaggage,
+    'handleOfferObj',
+    offerHandlerTaker => ({ offerHandlerTaker }),
+    {
+      handleOffer: ({ state }, invitationHandle, seatData) => {
+        const zcfSeat = seatManager.makeZCFSeat(seatData);
+        // TODO: provide a details that's a better diagnostic for the
+        // ephemeral offerHandler that did not survive upgrade.
+
+        const offerHandler = state.offerHandlerTaker.take(invitationHandle);
+        const offerResultP =
+          typeof offerHandler === 'function'
+            ? E(offerHandler)(zcfSeat, seatData.offerArgs)
+            : E(offerHandler).handle(zcfSeat, seatData.offerArgs);
+
+        const offerResultPromise = offerResultP.catch(reason => {
+          if (reason === undefined) {
+            const newErr = new Error(
+              `If an offerHandler throws, it must provide a reason of type Error, but the reason was undefined. Please fix the contract code to specify a reason for throwing.`,
+            );
+            throw zcfSeat.fail(newErr);
+          }
+          throw zcfSeat.fail(reason);
+        });
+        const exiter = makeExiter(seatData.proposal, zcfSeat);
+        /** @type {HandleOfferResult} */
+        return harden({ offerResultPromise, exitObj: exiter });
+      },
+    },
+  );
+
+  const HandleOfferFunctionShape = M.remotable('HandleOfferFunction');
+  const OfferHandlerShape = M.or(HandleOfferShape, HandleOfferFunctionShape);
+  const TakerI = M.interface('offer handler taker', {
+    take: M.call(InvitationHandleShape)
+      .optional(M.string())
+      .returns(OfferHandlerShape),
+  });
+  const taker = vivifyFarInstance(zcfBaggage, 'offer handler taker', TakerI, {
+    take: takeOfferHandler,
+  });
+  const handleOfferObj = makeHandleOfferObj(taker);
+
+  const evaluateContract = () => {
+    let bundle;
+    if (passStyleOf(contractBundleCap) === 'remotable') {
+      const bundleCap = contractBundleCap;
+      // @ts-expect-error vatPowers is not typed correctly: https://github.com/Agoric/agoric-sdk/issues/3239
+      bundle = powers.D(bundleCap).getBundle();
+    } else {
+      bundle = contractBundleCap;
+    }
+    return evalContractBundle(bundle);
+  };
+  // evaluate the contract (either the first version, or an upgrade)
+  const { start, buildRootObject, privateArgsShape, customTermsShape, vivify } =
```

Comment:
This is a user-visible change to the Zoe API, right?

I think we document `start` at https://docs.agoric.com/guides/zoe/#writing-and-installing-a-contract
I wonder if we document `vivify`.

`buildRootObject` is documented in https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/docs/bundles.md ; in this context, it seems to be used just for some nicer diagnostics.

Reply:
`vivify` is part of the durability package. @tyrosine has that high on his list to start on, but we don't have but scattered mentions of it in our docs, yet. 

> This is a user-visible change to the Zoe API, right?

it's a change, but no existing code is impacted.

Reply:
> `vivify` is part of the durability package. @tyrosine has that high on his list to start on

I hope we can use https://github.com/Agoric/documentation/issues for his list soon. I don't see `vivify` there (nor durable, durability).

> > This is a user-visible change to the Zoe API, right?
> 
> it's a change, but no existing code is impacted.

Right; I'm suggesting that the **Documentation considerations** section of this PR is missing something. And I guess I'm suggesting that if this is adopted, an issue should be opened in https://github.com/Agoric/documentation/issues .

Reply:
Good call. I've filed https://github.com/Agoric/documentation/issues/743 and referenced it